clients/src/main/java/org/apache/kafka/clients/ApiVersions.java

```diff
@@ -37,9 +37,13 @@ public class ApiVersions {
     private final Map<String, NodeApiVersions> nodeApiVersions = new HashMap<>();
     private byte maxUsableProduceMagic = RecordBatch.CURRENT_MAGIC_VALUE;
 
+    // The maximum finalized feature epoch of all the node api versions.
+    private long maxFinalizedFeaturesEpoch = -1;
+
     public synchronized void update(String nodeId, NodeApiVersions nodeApiVersions) {
         this.nodeApiVersions.put(nodeId, nodeApiVersions);
```

Comment:
Do we want to only update if the epoch is new?

Reply:
Done